mlflow/sklearn/__init__.py

```diff
@@ -932,9 +1390,46 @@ def patched_fit(original, self, *args, **kwargs):
         )
     ]
 
+    def should_patch_class_method(class_def, func_name):
+        try:
+            if hasattr(class_def, func_name):
+                method = getattr(class_def, func_name)
+                return callable(method)
+            return False
+        except:
+            return False
+
+    def patched_IffHasAttrDescriptor_get(self, obj, type=None):
+        from functools import update_wrapper
+        # raise an AttributeError if the attribute is not present on the object
+        if obj is not None:
+            # delegate only on instances, not the classes.
+            # this is to allow access to the docstrings.
+            for delegate_name in self.delegate_names:
+                try:
+                    delegate = sklearn.utils.metaestimators.attrgetter(delegate_name)(obj)
+                except AttributeError:
+                    continue
+                else:
+                    getattr(delegate, self.attribute_name)
+                    break
+            else:
+                sklearn.utils.metaestimators.attrgetter(self.delegate_names[-1])(obj)
+
+        # lambda, but not partial, allows help() to work with update_wrapper
+        if obj is None:  # `cls.method`
+            out = lambda *args, **kwargs: self.fn(*args, **kwargs)
+        else:  # `instance.method`
+            out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)
+        # update the docstring of the returned function
+        update_wrapper(out, self.fn)
+        return out
+
+    sklearn.utils.metaestimators._IffHasAttrDescriptor.__get__ = patched_IffHasAttrDescriptor_get
```

Comment:
Can we `safe_patch()` this in case our inserted logic fails?

Reply:
scikit-learn recently made changes on `_IffHasAttrDescriptor`:
https://github.com/scikit-learn/scikit-learn/pull/19948

Reply:
I found only patching `_IffHasAttrDescriptor.__get__` is not sufficient. We need also attach the `__get__` method (with the delegate behavior) to the patched method. The intention of the `_IffHasAttrDescriptor` is to delegate `hasattr` result to other objects. e.g. `hasattr(pipeline_instance, 'predict_proba')` may return `True` or `False`, it will be delegated to call `hasattr(pipeline_instance._final_estimator, 'predict_proba')`, but, our mlflow safe patching break this behavior and make `hasattr(pipeline_instance, 'predict_proba')` always return `True`.
Let's skip patching `if_delegate_has_method` decorated method in this PR and make complete fixing in follow-up PR.

CC @harupy @dbczumar